### PR TITLE
*:Record the time consuming of memory operation of Insert Executor in Runtime Information (#19574)

### LINKED ISSUE #19512: Record the time consuming of memory operation of Insert Executor in Runtime Information
## Development Task

INSERT IGNORE from a table of 300,000 rows,

```mysql
INSERT IGNORE INTO t SELECT * FROM t1
```
And we got the following slow query log.
image

Unfortunately, with this log, we have no clue where the time is consumed. 

Then we must investigate this problem by using go profile or trace.

image

With the above go profile graph, I propose to add two extra time-consuming sections in runtime information and trace span,
1. Check Duplicate, including [`getKeysNeedCheck`](https://github.com/pingcap/tidb/blob/650be7c4383a197a0311433572d082eb087217a2/executor/insert_common.go#L952) and [`txn.Get` from memory buffer](https://github.com/pingcap/tidb/blob/650be7c4383a197a0311433572d082eb087217a2/executor/insert_common.go#L978), but except the RPC time.
2. Add Records, for the memory operation of memtable.

## SIG slack channel
[#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score
- 300

## Mentor
 * @qw4990

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #19574 to release-4.0

---

<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

Issue Number: close #19512 
Problem Summary:

### What is changed and how it works?

What's Changed:
add two extra time-consuming sections in runtime information and trace span
How it Works:
Split inspection and insertion into two parts, and time each part
### Related changes

- PR to update /`pingcap/exector`:
- Need to cherry-pick to the release branch

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects
**After this pr**:
```
mysql> select tidb_decode_plan('8QLgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI2Mi4wNDjCtXMsIGxvb3BzOjEsIHByZXBhcmU6MzMuNDI2BRxMY2hlY2tfaW5zZXJ0Ont0b3RhbF8JQRQyOC42MjIFJQxtZW1fMikAGDE1MS4zMjEFHTRwcmVmZXRjaDo3Ny4zMAkUSHJwYzp7QmF0Y2hHZXQ6e251bV8BEwgxLCAuYgAMNS4xNwEwBH0sbiYAmDkuMzI3wrVzfX19LCBjb21taXRfdHhuOiB7cHJld3JpdGU6OTQuMQ16DGdldF8RJRRzOjQuOTQFpQkTWDoxLjMwNjU1OG1zLCByZWdpb25fbnVtAZgFSRxfa2V5czoyLA0OWGJ5dGU6NDh9CTE5MiBCeXRlcwlOL0EK');
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_decode_plan('8QLgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI2Mi4wNDjCtXMsIGxvb3BzOjEsIHByZXBhcmU6MzMuNDI2BRxMY2hlY2tfaW5zZXJ0Ont0b3RhbF8JQRQyOC42MjIFJQxtZW1fMikAGDE1MS4zMjEFHTRwcmVmZXRjaDo3Ny4zMAkUSHJwYzp7QmF0Y2hHZXQ6e251bV8BEwgxLCAuYgAMNS4xNwEwBH0sbiYAmDkuMzI3wr                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 	id      	task	estRows	operator info	actRows	execution info                                                                                                                                                                                                                                                                                                                                   	memory   	disk
	Insert_1	root	0      	N/A          	0      	time:262.048µs, loops:1, prepare:33.426µs, check_insert:{total_time:228.622µs, mem_check_insert:151.321µs, prefetch:77.301µs, rpc:{BatchGet:{num_rpc:1, total_time:25.171µs},Get:{num_rpc:1, total_time:29.327µs}}}, commit_txn: {prewrite:94.101µs, get_commit_ts:4.94µs, commit:1.306558ms, region_num:1, write_keys:2, write_byte:48}	192 BytesN/A          |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
```
mysql> select tidb_decode_plan('1AHgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI1MS44MzHCtXMsIGxvb3BzOjEsIHByZXBhcmU6NjQuOTM4BRxMY2hlY2tfaW5zZXJ0Ont0b3RhbF8FQRgxODYuODkzBSUMbWVtXzIpAAw3Mi4xDV08cHJlZmV0Y2g6MTE0Ljc2MgUxSHJwYzp7QmF0Y2hHZXQ6e251bV8BEwgxLCAdYmgzNy4wMDfCtXN9fX0JMTQ0IEJ5dGVzCU4vQQo=');
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_decode_plan('1AHgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI1MS44MzHCtXMsIGxvb3BzOjEsIHByZXBhcmU6NjQuOTM4BRxMY2hlY2tfaW5zZXJ0Ont0b3RhbF8FQRgxODYuODkzBSUMbWVtXzIpAAw3Mi4xDV08cHJlZmV0Y2g6MTE0Ljc2MgUxSHJwYzp7QmF0Y2hHZXQ6e251bV8BEwgxLCAdYmgzNy4wMDfCtXN9fX0JMTQ0IEJ5dG                                                                                                                                                                                                                                 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 	id      	task	estRows	operator info	actRows	execution info                                                                                                                                                                      	memory   	disk
	Insert_1	root	0      	N/A          	0      	time:251.831µs, loops:1, prepare:64.938µs, check_insert:{total_time:186.893µs, mem_check_insert:72.131µs, prefetch:114.762µs, rpc:{BatchGet:{num_rpc:1, total_time:37.007µs}}}	144 Bytes	N/A       |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
```
mysql> select tidb_decode_plan('1gHgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI3NS4zNDPCtXMsIGxvb3BzOjEsIHByZXBhcmU6OTguMjExBRw0aW5zZXJ0OjE3Ny4xMzIFE3Bjb21taXRfdHhuOiB7cHJld3JpdGU6MjY4LjQ2OAUiDGdldF8RJhRzOjMuOTMJGAkUGDozOTguMDcFKjRyZWdpb25fbnVtOjEsIAVKEF9rZXlzGQ4EYnkBXUA5fQkxNzMgQnl0ZXMJTi9BCg==');
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_decode_plan('1gHgMAkyN18xCTAJMAlOL0EJMAl0aW1lOjI3NS4zNDPCtXMsIGxvb3BzOjEsIHByZXBhcmU6OTguMjExBRw0aW5zZXJ0OjE3Ny4xMzIFE3Bjb21taXRfdHhuOiB7cHJld3JpdGU6MjY4LjQ2OAUiDGdldF8RJhRzOjMuOTMJGAkUGDozOTguMDcFKjRyZWdpb25fbnVtOjEsIAVKEF9rZXlzGQ4EYnkBXUA5fQkxNzMgQn                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 	id      	task	estRows	operator info	actRows	execution info                                                                                                                                                                        	memory   	disk
	Insert_1	root	0      	N/A          	0      	time:275.343µs, loops:1, prepare:98.211µs, insert:177.132µs, commit_txn: {prewrite:268.468µs, get_commit_ts:3.938µs, commit:398.07µs, region_num:1, write_keys:1, write_byte:29}	173 Bytes	N/A       |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
```
mysql> trace format='log' insert into t values(10,1) on duplicate key update `a`=`a`+1;
+----------------------------+------------------------------------------------------------------------+------+--------------------------------+
| time                       | event                                                                  | tags | spanName                       |
+----------------------------+------------------------------------------------------------------------+------+--------------------------------+
| 2020-09-17 19:24:08.424576 | --- start span trace ----                                              |      | trace                          |
| 2020-09-17 19:24:08.424587 | --- start span session.Execute ----                                    |      | session.Execute                |
| 2020-09-17 19:24:08.424592 | execute: insert into t values(10,1) on duplicate key update `a`=`a`+1  |      | session.Execute                |
| 2020-09-17 19:24:08.424593 | --- start span session.ParseSQL ----                                   |      | session.ParseSQL               |
| 2020-09-17 19:24:08.424610 | --- start span session.ExecuteStmt ----                                |      | session.ExecuteStmt            |
| 2020-09-17 19:24:08.424612 | --- start span executor.Compile ----                                   |      | executor.Compile               |
| 2020-09-17 19:24:08.424722 | --- start span session.runStmt ----                                    |      | session.runStmt                |
| 2020-09-17 19:24:08.424749 | --- start span executor.handleNoDelayExecutor ----                     |      | executor.handleNoDelayExecutor |
| 2020-09-17 19:24:08.424752 | --- start span *executor.InsertExec.Next ----                          |      | *executor.InsertExec.Next      |
| 2020-09-17 19:24:08.424783 | insert 1 rows into table `t`                                           |      | *executor.InsertExec.Next      |
| 2020-09-17 19:24:08.424829 | --- start span prefetchDataCache ----                                  |      | prefetchDataCache              |
| 2020-09-17 19:24:08.424836 | --- start span prefetchUniqueIndices ----                              |      | prefetchUniqueIndices          |
| 2020-09-17 19:24:08.424841 | --- start span tikvTxn.BatchGet ----                                   |      | tikvTxn.BatchGet               |
| 2020-09-17 19:24:08.424876 | load region 66 from pd, due to cache-miss                              |      | tikvTxn.BatchGet               |
| 2020-09-17 19:24:08.424882 | --- start span regionReqauest.SendReqCtx ----                          |      | regionReqauest.SendReqCtx      |
| 2020-09-17 19:24:08.424890 | send BatchGet request to region 66 at store1                           |      | regionReqauest.SendReqCtx      |
| 2020-09-17 19:24:08.424918 | --- start span prefetchConflictedOldRows ----                          |      | prefetchConflictedOldRows      |
| 2020-09-17 19:24:08.424919 | --- start span tikvTxn.BatchGet ----                                   |      | tikvTxn.BatchGet               |
| 2020-09-17 19:24:08.424947 | --- start span executor.updateRecord ----                              |      | executor.updateRecord          |
| 2020-09-17 19:24:08.424981 | --- start span tikvSnapshot.get ----                                   |      | tikvSnapshot.get               |
| 2020-09-17 19:24:08.424984 | --- start span regionReqauest.SendReqCtx ----                          |      | regionReqauest.SendReqCtx      |
| 2020-09-17 19:24:08.424986 | send Get request to region 66 at store1                                |      | regionReqauest.SendReqCtx      |
| 2020-09-17 19:24:08.425057 | --- start span session.RollbackTxn ----                                |      | session.RollbackTxn            |
| 2020-09-17 19:24:08.425108 | execute with error(0): [kv:1062]Duplicate entry '11' for key 'PRIMARY' |      | trace                          |
+----------------------------+------------------------------------------------------------------------+------+--------------------------------+
24 rows in set, 1 warning (0.00 sec)
```
- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note <!-- bugfixes or new feature need a release note -->
- add two extra time-consuming sections in runtime information and trace span<!-- Please write a release note here to describe the change you made when it is released to the users of TiDB. If your PR doesn't involve any change to TiDB(like test enhancements, RFC proposals...), you can write `No release note`. -->

### CODE REVIEW

File: executor/insert_common.go [hunk 1]
```diff
@@ -929,12 +938,37 @@ func (e *InsertValues) handleWarning(err error) {
 	sc.AppendWarning(err)
 }
 
+<<<<<<< HEAD
```
Comment (qw4990) [inline executor/insert_common.go:941]: @jyz0309 Please fix conflicts.

**Review by crazycs520 — COMMENTED** [review]
LGTM

**Comment by crazycs520** [comment]
@qw4990 PTAL

**Review by qw4990 — COMMENTED** [review]
LGTM

**Review by qw4990 — APPROVED** [review]

**Review by SunRunAway — APPROVED** [review]

**Pull request merged**